### PR TITLE
k_object/syscall: cleanup internal headers

### PR DESCRIPTION
**Pull request by nashif**
This is a proposal to cleanup internal APIs and document them properly and decouple them from private APIs using the z/Z prefix.

The internal APIs are moved to an internal header folder and prefixes are changed.
Documentation of many of those APIs is also adapted to indicate they are internal APIs and they doxygen output is put in a special section alongside other internal APIs 

- kobject: move internal functions to own header
- kernel: objects: rename z_dynamic_object_create -> k_object_create_dynamic
- kernel: objects: rename z_dynamic_object_aligned_create
- kernel: object: rename z_object_init to k_object_init
- kernel: object: rename z_object -> k_object
- kernel: move syscall_handler.h to internal include directory
- k_is_in_user_syscall
- syscall: Z_SYSCALL_VERIFY_MSG -> K_SYSCALL_VERIFY_MSG
- syscall: Z_SYSCALL_IS_OBJ ->  K_SYSCALL_IS_OBJ
- syscall: Z_SYSCALL_MEMORY_ARRAY -> K_SYSCALL_MEMORY_ARRAY
- syscall: z_obj_validation_check -> k_object_validation_check
- syscall: z_object_recycle -> k_object_recycle
- syscall: z_object_uninit -> k_object_uninit
- syscall: z_thread_perms_all_clear -> k_thread_perms_all_clear
- z_thread_perms_clear -> k_thread_perms_clear
- syscall: z_thread_perms_set -> k_thread_perms_set
- syscall: z_thread_perms_inherit -> k_thread_perms_inherit
- syscall: z_object_wordlist_foreach -> k_object_wordlist_foreach
- syscall: z_object_find -> k_object_find
- syscall: z_object_validate -> k_object_validate
- syscall: z_dump_object_error -> k_object_dump_error
- syscall: z_user_string_nlen -> k_usermode_string_nlen
- syscall: z_user_alloc_from_copy -> k_usermode_alloc_from_copy
- syscall: z_user_from_copy -> k_usermode_from_copy
- syscall: rename z_user_ to k_usermode_
- syscall: rename Z_SYSCALL_ to K_SYSCALL_
- syscall: rename Z_SYSCALL_DRIVER_ -> K_SYSCALL_DRIVER_
- syscall: rename Z_SYSCALL_VERIFY -> K_SYSCALL_VERIFY
- syscall: rename Z_OOPS -> K_OOPS
- doc: define internal API group
- syscall/userspace: doc: include in doxygen
- syscall: add a note about the API being internal
- arch: move arch interface to internal api group
- doc

### CODE REVIEW

**Comment by nashif** [comment]
Where the internal docs will show up:
https://builds.zephyrproject.io/zephyr/pr/63153/docs/doxygen/html/group__internal__api.html

This group of important but internal APIs was never part of the documentation, now we have this in doxygen and available to be referenced in RST.

https://builds.zephyrproject.io/zephyr/pr/63153/docs/doxygen/html/group__syscall__apis.html

A note was added about the APIs being internal.

**Review by dcpleung — COMMENTED** [review]
Has the community figured out how to name private functions? The gperf and permission getting/setting functions in kernel should only be used by the kernel.

File: include/zephyr/internal/syscall_handler.h [hunk 1]
```diff
@@ -48,6 +48,9 @@ enum _obj_init_check {
  * user thread. If the system call was invoked from supervisor mode,
  * or we are not handling a system call, this will return false.
  *
+ * @note This is an internal API. Do not use unless you are extending
```
[the remark this answers is not in the document]
Comment (nashif) [reply include/zephyr/internal/syscall_handler.h:51]: yeah, will be adding an alias I guess.

**Review by de-nordic — DISMISSED** [review]
Flash API looks ok.

**Review by dcpleung — DISMISSED** [review]

File: doc/releases/release-notes-2.2.rst [hunk 2]
```diff
@@ -198,7 +198,7 @@ Kernel
   has already been completed
 * Disable preemption when handing fatal errors
 * Fix an issue with the sytsem call stack frame if the system call is
-  preempted and then later tries to Z_OOPS()
+  preempted and then later tries to K_OOPS()
```
Comment (jhedberg) [inline doc/releases/release-notes-2.2.rst:201]: We should probably not modify old release notes

File: doc/releases/release-notes-2.3.rst [hunk 3]
```diff
@@ -1350,7 +1350,7 @@ release:
 * :github:`17888` - arch/x86: remove IAMCU ABI support
 * :github:`17775` - Microchip XEC rtos timer should be using values coming from DTS
 * :github:`17755` - ARC privilege mode stacks waste memory due to alignment requirements
-* :github:`17735` - abolish Z_OOPS() in system call handlers
+* :github:`17735` - abolish K_OOPS() in system call handlers
```
Comment (jhedberg) [inline doc/releases/release-notes-2.3.rst:1353]: Same here

**Review by jhedberg — CHANGES_REQUESTED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**